### ISSUE TITLE
Rustdoc does not substitute const parameters

### ISSUE THREAD

**Issue by fmease** [rust-lang/rust#82852]
I ran `cargo doc` on this code (`src/lib.rs`):

```rust
#![feature(min_const_generics)]

type Array<T, const N: usize> = [T; N];

pub type A0 = [i32; 3];
pub type A1 = Array<i32, 3>;
pub struct B0(pub [i32; 3]);
pub struct B1(pub Array<i32, 3>);
```

I expected to see the definitions `type A1 = [i32; 3];` and `pub struct B1(pub [i32; 3]);` in the generated documentation.
Instead, I found `type A1 = [i32; N];` and `pub struct B1(pub [i32; N]);`.

That is, the definitions wrongly contain a stray identifier `N` that should have been substituted with the literal `3` just like the `T` was successfully substituted with `i32`.

### Meta

`rustc --version --verbose`:
```
rustc 1.52.0-nightly (51748a8fc 2021-03-05)
binary: rustc
commit-hash: 51748a8fc77283914d4135f31b5966a407208187
commit-date: 2021-03-05
host: x86_64-unknown-linux-gnu
release: 1.52.0-nightly
LLVM version: 12.0.0
```

@rustbot modify labels: +T-rustdoc +A-const-generics +F-min_const_generics +C-bug

**Comment by fmease** [comment]
I hope this is not a duplicate or an already known issue that is part of some bigger tracking issue or some agenda. I thoroughly searched beforehand but couldn't find anything related.

**Comment by jyn514** [comment]
This is a duplicate of https://github.com/rust-lang/rust/issues/14072 I think.

**Comment by fmease** [comment]
Hmm, maybe it is. Hard to tell as it is really vague and updates to this old issue are scattered.

In my eyes, this simple case of const parameters (literals only) looks manageable to implement. Not so sure about the general case which requires const evaluation. @ jyn514, do you think writing the code for the substitution of const parameters is a matter of "copying"/mirroring the one for the substitution of type parameters (which in some parts has to exist as seen by `T` -> `i32`)? Thinking about a potential first contribution.

**Comment by jyn514** [comment]
> In my eyes, this simple case of const parameters (literals only) looks manageable to implement. Not so sure about the general case which requires const evaluation.

You can evaluate constants with [`print_evaluated_const`](https://doc.rust-lang.org/nightly/nightly-rustc/rustdoc/clean/utils/fn.print_evaluated_const.html): https://github.com/rust-lang/rust/blob/27885a94c6a6c98a0d49af3962f1218adcd6efd2/src/librustdoc/clean/mod.rs#L1957

> @ jyn514, do you think writing the code for the substitution of const parameters is a matter of "copying"/mirroring the one for the substitution of type parameters (which in some parts has to exist as seen by T -> i32)? Thinking about a potential first-time contribution.

I'm not sure how this part of the code works, sorry. Happy to look into it if you get stuck though :)

**Comment by fmease** [comment]
> Happy to look into it if you get stuck though :)

Thank you :) I appreciate that. I opened a [Zulip stream](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/Getting.20to.20know.20the.20code.20base.20to.20fix.20.2382852/near/229476331) to discuss the work on the implementation.

**Comment by clarfonthey** [comment]
Provided another example in a (now closed) duplicate issue:

```rust
pub trait Test {}
pub struct Array<T, const N: usize>([T; N]);

impl Test for [u8; 1 + 1] {}
impl Test for Array<u8, {1 + 1}> {}
```

![screenshot of docs](https://user-images.githubusercontent.com/15850505/116953578-a11d8580-ac5b-11eb-88bc-d602c6c138cf.png)

**Comment by fmease** [comment]
@rustbot modify labels: -F-min_const_generics
(the feature has been stabilized)

**Comment by camelid** [comment]
I think the issue here is that `DocContext.ct_substs` is actually never read from. I'm looking into this a bit.

**Comment by fmease** [comment]
@camelid Yes, I made the same observation and noted down some more findings in the Zulip stream linked above (although some information there might be outdated by now).
